### PR TITLE
Refactoring hash code

### PR DESCRIPTION
**Pull request by yilongli**
I am creating this pull request in order to test it. Besides, I will wait for Andrei's soon-coming pull request to get merged first.

### CODE REVIEW

**Comment by yilongli** [comment]
@andreistefanescu @dwightguth please review

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Term.java [hunk 1]
```diff
@@ -181,4 +182,22 @@ public Term substituteWithBinders(Variable variable, Term term, TermContext cont
...
     }
+    
+    /**
+     * Computes and caches the hashCode if it has not been computed yet.
+     * Otherwise, simply returns the cached value.
+     */
+    @Override
+    public final int hashCode() {
+        if (hashCode == Utils.NO_HASHCODE) {
+            hashCode = computeHash();
```
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Term.java:193]: I believe this was a suggestion of @dwightguth: if computeHash() returns 0, set the hashCode to 1.

**Comment by andreistefanescu** [comment]
lgtm

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java [hunk 2]
```diff
@@ -83,7 +78,7 @@ public final boolean isSymbolic() {
     }
     
     @Override
-    public int hashCode() {
+    public int computeHash() {
         throw new UnsupportedOperationException();
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java]: make this abstract?

**Comment by dwightguth** [comment]
looks good other than the two comments I and Andrei made.

**Pull request merged**